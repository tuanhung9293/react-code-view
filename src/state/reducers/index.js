import {combineReducers} from 'redux'
import collections from './collections'
import medias from './medias'
import forms from './forms'
import ui from './ui'

const root = combineReducers({
  collections,
  medias,
  forms,
  ui
})

function rootReducer(state, action) {
  return root(state, action)
}

export default rootReducer