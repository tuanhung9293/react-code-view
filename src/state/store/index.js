import {createStore, applyMiddleware, compose} from 'redux'
import thunk from 'redux-thunk'
import {persistStore, persistReducer} from 'redux-persist'
import storage from 'redux-persist/lib/storage'
import trackWindowMiddleware from 'state/middleware/trackWindowMiddleware'
import rootReducer from '../reducers'
import {startTrackingWindow} from '../actions/uiAction'

const middlewares = [
  thunk,
  trackWindowMiddleware
]

const persistConfig = {
  key: 'root',
  storage: storage,
  whitelist: ['collections']
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

const isDev = process.env.NODE_ENV === 'development'

const composeEnhancers =
  (isDev && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose

export function createBaseStore(initialState = {}) {
  return createStore(
    persistedReducer,
    {
      ...initialState
    },
    composeEnhancers(applyMiddleware(...middlewares))
  )
}

const store = createBaseStore()
store.dispatch(startTrackingWindow())

const persistor = persistStore(store)

export {store, persistor}
