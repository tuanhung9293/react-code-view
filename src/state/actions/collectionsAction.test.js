import {
  addCollection,
  editCollection,
  removeCollection,
  likeCollection,
  sortNasaCollections
} from './collectionsAction'
import {createBaseStore} from 'state/store'

const dateNowStub = jest.fn(() => 1530518207007)
global.Date.now = dateNowStub

describe('collectionsAction integration test', () => {
  it('addCollection add new collection', () => {
    const store = createBaseStore({
      collections: {
        data: [{nasaId: 12, value: 'value12'}]
      }
    })

    store.dispatch(addCollection({nasaId: 13, value: 'value13'}))
    expect(store.getState().collections).toMatchSnapshot()
  })

  it('addCollection add existing collection', () => {
    const store = createBaseStore({
      collections: {
        data: [{nasaId: 12, value: 'value12'}]
      }
    })

    store.dispatch(addCollection({nasaId: 12, value: 'value13'}))
    expect(store.getState().collections).toMatchSnapshot()
  })

  
  it('editCollection', () => {
    const store = createBaseStore({
      collections: {
        data: [{nasaId: 12, value: 'value12'}]
      }
    })

    store.dispatch(editCollection({nasaId: 12, value: 'value13'}))
    expect(store.getState().collections).toMatchSnapshot()
  })

  it('removeCollection', () => {
    const store = createBaseStore({
      collections: {
        data: [{nasaId: 12, value: 'value12'}, {nasaId: 14, value: 'value14'}]
      }
    })

    store.dispatch(removeCollection(12))
    expect(store.getState().collections).toMatchSnapshot()
  })

  it('likeCollection', () => {
    const store = createBaseStore({
      collections: {
        data: [{nasaId: 12, value: 'value12', liked: false}, {nasaId: 14, value: 'value14', liked: true}]
      }
    })

    store.dispatch(likeCollection(12))
    expect(store.getState().collections).toMatchSnapshot()
  })

  it('sortNasaCollections by title decrease', () => {
    const store = createBaseStore({
      collections: {
        data: [
          {nasaId: 12, value: 'value12', title: 'a'},
          {nasaId: 14, value: 'value14', title: 'c'},
          {nasaId: 14, value: 'value14', title: 'b'}
        ]
      }
    })

    store.dispatch(sortNasaCollections('title,1'))
    expect(store.getState().collections).toMatchSnapshot()
  })
  
  it('sortNasaCollections by dateCreated increase', () => {
    const store = createBaseStore({
      collections: {
        data: [
          {nasaId: 12, value: 'value12', dateCreated: "2018-11-16T00:00:00Z"},
          {nasaId: 14, value: 'value14', dateCreated: "2018-12-16T00:00:00Z"},
          {nasaId: 14, value: 'value14', dateCreated: "2018-10-16T00:00:00Z"}
        ]
      }
    })

    store.dispatch(sortNasaCollections('dateCreated,-1'))
    expect(store.getState().collections).toMatchSnapshot()
  })
})
  