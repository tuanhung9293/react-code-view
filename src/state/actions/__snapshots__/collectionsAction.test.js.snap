// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectionsAction integration test addCollection add existing collection 1`] = `
Object {
  "data": Array [
    Object {
      "dateEdited": 1530518207007,
      "nasaId": 12,
      "value": "value13",
    },
  ],
}
`;

exports[`collectionsAction integration test addCollection add new collection 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "dateEdited": 1530518207007,
      "nasaId": 13,
      "value": "value13",
    },
  ],
}
`;

exports[`collectionsAction integration test editCollection 1`] = `
Object {
  "data": Array [
    Object {
      "dateEdited": 1530518207007,
      "nasaId": 12,
      "value": "value13",
    },
  ],
}
`;

exports[`collectionsAction integration test likeCollection 1`] = `
Object {
  "data": Array [
    Object {
      "liked": true,
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "liked": true,
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collectionsAction integration test removeCollection 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collectionsAction integration test sortNasaCollections by dateCreated increase 1`] = `
Object {
  "data": Array [
    Object {
      "dateCreated": "2018-10-16T00:00:00Z",
      "nasaId": 14,
      "value": "value14",
    },
    Object {
      "dateCreated": "2018-11-16T00:00:00Z",
      "nasaId": 12,
      "value": "value12",
    },
    Object {
      "dateCreated": "2018-12-16T00:00:00Z",
      "nasaId": 14,
      "value": "value14",
    },
  ],
}
`;

exports[`collectionsAction integration test sortNasaCollections by title decrease 1`] = `
Object {
  "data": Array [
    Object {
      "nasaId": 14,
      "title": "c",
      "value": "value14",
    },
    Object {
      "nasaId": 14,
      "title": "b",
      "value": "value14",
    },
    Object {
      "nasaId": 12,
      "title": "a",
      "value": "value12",
    },
  ],
}
`;
